light-client/relayer/tests/block_processing.rs

```diff
@@ -0,0 +1,191 @@
+    let ledger_db_path = {
+        let ledger_db_path = blockchain_path.join("ledger_db");
+        std::fs::create_dir(&ledger_db_path).expect("failed to create ledger dir");
+        LedgerDB::create(&ledger_db_path).expect("failed to create LedgerDB");
+        ledger_db_path
+    };
+
+    let mut ledger = LedgerDB::open(&ledger_db_path).unwrap();
+    let tx_source_urls: Vec<Url> = (0..5)
+        .filter_map(|idx| Url::parse(&format!("https://node{}.localhost", idx)).ok())
+        .collect();
+
+    let watcher = WatcherDB::open_rw(&watcher_db_path, &tx_source_urls, logger.clone())
+        .expect("Could not create watcher_db");
+
+    // Initialize ledger
+    let sender = AccountKey::random(&mut rng);
+    let num_blocks = 1;
+    initialize_ledger(BLOCK_VERSION, &mut ledger, num_blocks, &sender, &mut rng);
+
+    // Make node signer identities
+    let signers: Vec<Ed25519Pair> = (0..5).map(|_| Ed25519Pair::from_random(&mut rng)).collect();
+
+    let config = Config {
+        start_block_index: 1,
+        min_signatures: 4,
+        ledger_db: ledger_db_path,
+        watcher_db: watcher_db_path,
+        admin_listen_uri: None,
+    };
+
+    log::info!(logger, "Starting relayer");
+
+    let runner = Relayer::new(
+        config,
+        ledger.clone(),
+        watcher.clone(),
+        DummySender {
+            logger: logger.clone(),
+        },
+        logger.clone(),
+    );
+
+    // Do happy path testing where there is a burn txo
+    let burn_txo = TxOut::new_with_memo(
+        BLOCK_VERSION,
+        Amount::new(100, 2.into()),
+        &burn_address(),
+        &FromRandom::from_random(&mut rng),
+        EncryptedFogHint::fake_onetime_hint(&mut rng),
+        |_ctxt| Ok(BurnRedemptionMemo::new([7u8; 64]).into()),
+    )
+    .unwrap(); // TODO: Use real format here?
+
+    let random_txo = TxOut::new(
+        BLOCK_VERSION,
+        Amount::new(100, 2.into()),
+        &PublicAddress::from_random(&mut rng),
+        &FromRandom::from_random(&mut rng),
+        EncryptedFogHint::fake_onetime_hint(&mut rng),
+    )
+    .unwrap();
+
+    log::info!(logger, "Adding block");
+
+    let (block, block_contents, block_metadata) =
+        append_tx_outs_as_block(&mut ledger, &[burn_txo.clone(), random_txo], signers);
+
+    log::info!(logger, "Adding block to watcher");
+
+    // Add this stuff to the watcher
+    for (idx, block_metadata) in block_metadata.iter().enumerate() {
+        let block_data = BlockData::new(
+            block.clone(),
+            block_contents.clone(),
+            None,
+            Some(block_metadata.clone()),
+        );
+
+        let url = Url::from_str(&format!("https://node{}.localhost", idx)).unwrap();
+
+        watcher.add_block_data(&url, &block_data).unwrap();
+    }
+
+    let mut retries = 100;
+    loop {
+        if retries == 0 {
+            panic!("relayer message not received");
+        }
+        retries -= 1;
+        let records = runner.get_burned_tx_records();
```

Comment:
:thought_balloon: I think this test follows other testing patterns in the code base so is ok as is.  This is mostly me thinking out loud with some food for thought moving forward.

There is a lot of test setup here (and in similar tests). 
The setup is mainly to create a blockchain, ledger DB and watcher DB.  We should have sufficient tests elsewhere around ledger and watcher being able to read from the blockchain.

The asserts in this test are checking that the burn block was correctly captured in the `get_burned_tx_records()`. I wonder if unit tests in the relayer which are able to throw sequences of blocks at it.

The interesting tests at the level of this file is how the error handling is done in `relayer.rs`. When the ledger DB says notfound or returns an error, same with watcher db.